### ISSUE TITLE
Why Sec-FedCM-CSRF and not Sec-Fetch-Mode

### ISSUE THREAD

**Issue by martinthomson** [w3c-fedid/FedCM#320]
[Sec-Fetch-Mode](https://w3c.github.io/webappsec-fetch-metadata/#sec-fetch-mode-header) seems purpose-built for this sort of thing.  Adding another header field doesn't really help a lot.

(A server will naturally ignore either a new Sec-Fetch-Mode value or the Sec-FedCM-CSRF thing.  The value of the former is that it will compress better and it reuses an existing mechanism.

**Comment by npm1** [comment]
Is the proposal to add a new mode for `Sec-Fetch-Mode`, which would be added to all browser requests within the FedCM API? That seems reasonable to me.

**Comment by cbiesinger** [comment]
Hmm yeah looking at https://fetch.spec.whatwg.org/#concept-request-mode it seems like adding a fedcm method might be best. @annevk do you have any thoughts on that?

**Comment by domenic** [comment]
I'm not sure about this. The fetch mode concept is currently used to figure out major modes of the processing model for fetch, in terms of what results in network errors, what kind of security properties are required, how credentials are processed, whether early hints are respected, whether responses are opaque, etc.

I can't find anywhere in the spec that actually invokes fetch in a rigorous way, e.g. says what request input is used to the fetch algorithm (this seems very bad!!). But I am hoping that the fetch that this spec does is just a normal "cors"-mode fetch. So inventing a new mode doesn't seem right, in that case.

**Comment by npm1** [comment]
Hmm thanks for the context. @martinthomson based on the above I suspect we should keep the new header field? WDYT?

**Comment by bvandersloot-mozilla** [comment]
> I can't find anywhere in the spec that actually invokes fetch in a rigorous way,

[This was one of Anne's complaints as well](https://github.com/fedidcg/FedCM/issues/261). I think that it started to get defined in sections like [check the root manifest](https://fedidcg.github.io/FedCM/#check-the-root-manifest), but isn't quite finished.

This spec doesn't actually do a normal "cors"-mode fetch- it has two types of request that are more specific. One is credential-less and referrer-less and another is credentialed. 

One way forward that makes sense to me is making two new Sec-Fetch-Mode values, "fedcm-credentialless" and "fedcm-credentialed", which each specify the correct behavior within the constraints of Fetch. This also removes the need for the Sec-FedCM-CSRF header.

**Comment by domenic** [comment]
I hope neither of those proposed fetches bypass CORS, and I hope even more ardently that neither of them rise to the level of behavioral differences that would require an entirely new fetch mode!!

**Comment by bvandersloot-mozilla** [comment]
> Instead, it would make most sense if they were normal, "cors"-mode fetches, which set the referrer policy and credentials mode fields of the request appropriately.

Okay, that would work. Then we do still need the Sec-FedCM-CSRF header.

**Comment by martinthomson** [comment]
Looking at this more closely, I don't think that the header is needed.

A fetch with the "cors" mode ensures that the content of a resource is not readable to a random site that invokes `fetch()` unless the (IDP) site deliberately enables access.  To ensure that only the browser reads the response, the site needs to do exactly nothing.  The site can use fetch headers to perform checks, and some amount of defense-in-depth is sensible here[^1], but the extra check is redundant.

For requests that mutate state (logout, token acquisition), we can insist upon a preflight by explicitly setting the [use-CORS-preflight flag](https://fetch.spec.whatwg.org/#use-cors-preflight-flag).  But that doesn't get us what is needed.  As far as I can tell, this is where things get a little sketchy because sites can set a referrer policy that prevents the inclusion of an Origin[^2].  So a missing Origin is not a strong signal that the request comes from the UA.

An alternative is to use the CSRF token defense used by most sites: include a value in an earlier request that proves that the request is genuine.

In all cases, this only requires that the endpoint that the browser interacts have a URL with sufficient high entropy as to be sufficient hard to guess by a random site that invokes fetch[^3].  I might even claim that this provides better defense than a passive signal like Sec-FedCM-CSRF because it doesn't require an additional check, though I guess that you could build an endpoint that doesn't pay any attention to a high-entropy URL component.

[^1]: You probably want to add `X-Content-Type-Options: nosniff` to the response.  Then check `Origin` and `Sec-Fetch-Dest` and so forth.

[^2]: I can't see why we would allow script to make a cross-site request without `Origin`, so I might be missing something: the fetch spec is, as always, inscrutable.

[^3]: The simplest way to implement something like this is to have a per-user URL, with the URL encoding the identity of the user, a nonce, and an authentication tag; see also <https://www.w3.org/TR/capability-urls/>.  Add a key identifier or version number and you can use the same symmetric key for many users.

**Comment by cbiesinger** [comment]
How would you provide an endpoint with a high-entropy URL, given that sites can just read it using a server-side proxying script?

@domenic since all of these fetches are browser-mediated and also we don't expose info to the website unless the user gives permission (by selecting an account), we do bypass CORS

**Comment by cbiesinger** [comment]
Also, in practice, even if we did use CORS, I'm sure IDPs will have to give access to everyone anyway since there's lots of IDPs and I doubt they want to run their CORS checks against the list every time.

**Comment by domenic** [comment]
Bypassing CORS is VERY BAD and MUST NOT BE DONE for new features.

**Comment by martinthomson** [comment]
I don't want this to seem facetious, but we bypass CORS for DoH.  That said, I agree with @domenic in this case: this can use CORS.

**Comment by cbiesinger** [comment]
Well, that is a very strong statement. We can probably support CORS, but why do you say that not using it is so bad in this case?

**Comment by cbiesinger** [comment]
Also, should we require CORS even for the uncredentialed requests?

**Comment by cbiesinger** [comment]
To be honest, requiring CORS would make security worse here IMO because then the accounts endpoint becomes world-readable from websites instead of only to the request mediated by the UA.

**Comment by martinthomson** [comment]
> requiring CORS [...] the accounts endpoint becomes world-readable

With CORS, the resource decides what is and isn't readable.  By default (except under some narrow conditions), content is not readable by any site.  The browser can see it, but it won't pass that information on unless the site explicitly permits it.

**Comment by yi-gu** [comment]
We don't want the IDP to know about the RP when fetching the accounts so at least CORS cannot be used there. In addition, since the "*" wildcard cannot be specified when responding to a credentialed requests request, the token request cannot use CORS, right? Otherwise the IDP must specify ALL the RP origins.

More importantly, there are 3 special things about FedCM API:
1. the credentialed response is not visible to the RP's process with site isolation
2. the URLs are not controlled by the RP (potential attacker), but are described in the IDP's manifest. i.e. they are fully controlled by the potential victim.
3. it's true that the UI is triggered by a script running on the RP, however the UI is closer to the IDP than to the RP w.r.t. source and access (especially before user granting permission). e.g. when fetching the user picture from the IDP to display on the UI, the CSP img-src directive on the RP should not be applied because it's never exposed / used on the RP.

Therefore I think the current "Sec-" header without CORS solution makes more sense in this case.

**Comment by domenic** [comment]
I'd encourage you to involve the security teams at your browsers if you are thinking of introducing another hole in the same origin model by bypassing CORS. We've previously gotten consensus across the web community that such violations are not acceptable, and I'm disappointed to see people trying to bypass it yet again. I don't really have the energy to continue litigating these issues, so I am just going to hope that such features get blocked in per-browser security review and cross-browser TAG review, in accordance with previous resolutions.

**Comment by bvandersloot-mozilla** [comment]
> We don't want the IDP to know about the RP when fetching the accounts so at least CORS cannot be used there. In addition, since the "*" wildcard cannot be specified when responding to a credentialed requests request, the token request cannot use CORS, right? Otherwise the IDP must specify ALL the RP origins.

CORS preflight requests are sent with Referer-Policy "same-origin" which seems acceptable. 
The `Origin` header will only be sent if the [response tainting](https://fetch.spec.whatwg.org/#concept-request-response-tainting) gets set to "cors", which I think we can manage to avoid for our non-credentialed requests.
Also, `Access-Control-Allow-Origin` may be taken from the `Origin` request header. 

I think it is entirely possible for this request to be CORS-protected, and it makes sense to use it since the result does  end up in the RP's content process.

Scrolling back, I think I agree with Martin that we don't even need the Sec-FedCM-CSRF header. We already have a high-entropy component in the request to fetch a token (nonce), so all of that discussion of high-entropy URL components is moot. What is the property we are defending by making sure the endpoints are only hit by the browser API? I can't come up with a good answer, but I admit I haven't gone to the scrollback of the original inclusion.

**Comment by bvandersloot-mozilla** [comment]
> CORS preflight requests are sent with Referer-Policy "same-origin" which seems acceptable. The `Origin` header will only be sent if the [response tainting](https://fetch.spec.whatwg.org/#concept-request-response-tainting) gets set to "cors", which I think we can manage to avoid for our non-credentialed requests. 

Looking more closely, I don't think this is actually true. It looks like our non-credentialed requests would hit [this fallthrough case](https://fetch.spec.whatwg.org/#ref-for-concept-request-response-tainting%E2%91%A7).

**Comment by annevk** [comment]
`-Mode` indeed seems wrongish, but what about `-Dest`? This does seem like it might warrant a new destination type.

**Comment by bvandersloot-mozilla** [comment]
> It looks like our non-credentialed requests would hit [this fallthrough case](https://fetch.spec.whatwg.org/#ref-for-concept-request-response-tainting%E2%91%A7). Unless there is a mechanism for suppressing the `Origin` header in CORS that I am missing, this is bad news.

Continuing the conversation with myself. I think this is fine if we perform the uncredentialed fetches with a de-priveleged client: one whose origin is opaque. This would set the Origin header to `null` for those requests and be fetch-spec compatible. We have the idea of a Null security principal in Gecko that is useful for de-privileging. 



> `-Mode` indeed seems wrongish, but what about `-Dest`? This does seem like it might warrant a new destination type.

I like `-Dest` the most of the proposed options (`-Dest`, `-Mode`, and `Sec-FedCM-CSRF`). This also adds the ability for a developer to only allow "credential" requests and not arbitrary fetches using their CSP.

**Comment by domfarolino** [comment]
Drive-by:

> I think this is fine if we perform the uncredentialed fetches with a de-priveleged client: one whose origin is opaque. This would set the Origin header to null for those requests and be fetch-spec compatible.

Out of curiosity, what would be the need for nulling out the `Origin` for the uncredentialed request? IIUC, this is the request with which we're specifically OK with exposing the RP to the IDP ([via the client id](https://fedidcg.github.io/FedCM/#idp-api-client-id-metadata-endpoint)), right? So is there a harm in keeping the origin too? Does that expose any more information?

**Comment by npm1** [comment]
I agree that there should not be a problem with the uncredentialed requests here, and I think they can use regular fetch. The complicated one is the credentialed one, which is one where we might need fetch expert help. I don't think doing a regular CORS fetch (with `Origin` being the RP) would work. As Christian pointed out, that would give the RPs the power to do those fetches outside of FedCM and read the accounts information.

**Comment by bvandersloot-mozilla** [comment]
> Drive-by:
> 
> > I think this is fine if we perform the uncredentialed fetches with a de-priveleged client: one whose origin is opaque. This would set the Origin header to null for those requests and be fetch-spec compatible.
> 
> Out of curiosity, what would be the need for nulling out the `Origin` for the uncredentialed request? IIUC, this is the request with which we're specifically OK with exposing the RP to the IDP ([via the client id](https://fedidcg.github.io/FedCM/#idp-api-client-id-metadata-endpoint)), right? So is there a harm in keeping the origin too? Does that expose any more information?

I was specifically talking about the requests for the [manifests](https://fedidcg.github.io/FedCM/#idp-api-manifest), which should not expose the RP to the IDP.

**Comment by bvandersloot-mozilla** [comment]
> I agree that there should not be a problem with the uncredentialed requests here, and I think they can use regular fetch. The complicated one is the credentialed one, which is one where we might need fetch expert help. I don't think doing a regular CORS fetch (with `Origin` being the RP) would work. As Christian pointed out, that would give the RPs the power to do those fetches outside of FedCM and read the accounts information.

@npm1: That makes a lot of sense. Thanks for the clarification. If we add a new Sec-fetch-dest value of "credential", the IDP can validate that header in the request with an identical guarantee that the client is not initiating the request except through this API. And it has the added benefit of integrating into the Content Security Policy so the RP can make restrictive policy decisions as well.

**Comment by johannhof** [comment]
The point was raised in https://github.com/fedidcg/FedCM/issues/320#issuecomment-1220163833 and I think it's central to this discussion. As to my knowledge, CORS in order to work requires the server to have knowledge of the request `Origin`. A central privacy guarantee in FedCM for at least one of the credentialed requests is that it happens without any association to the originating site (i.e. with a `no-referrer` policy).

**Comment by annevk** [comment]
Not necessarily, if the origin is an opaque origin, you could respond with \``null`\`, even with credentialed fetches. Normally you wouldn't really do that, but perhaps for a dedicated endpoint it's okay.

**Comment by cbiesinger** [comment]
With regard to @domenic 's request to talk to the browser's security team, I want to be clear that we did do that and they were explicitly OK with not using CORS.

I think my comment that using CORS would make the accounts endpoint/token endpoint was misunderstood, I want to elaborate:
- IDPs have lots of RPs they support
- Therefore, they can't list all of them in an allowed origins header
- Therefore, I expect IDPs will have to allow "*" (or echo back the origin we send to the server)
- Therefore, the accounts and token endpoint will be world readable, if we require CORS for FedCM
- This could be fixed by IDPs checking Sec-FedCM-CSRF or Sec-Fetch-Dest but since the failure case is so bad it would be better if we did not have to rely on that

**Comment by johannhof** [comment]
> Not necessarily, if the origin is an opaque origin, you could respond with `null`, even with credentialed fetches. Normally you wouldn't really do that, but perhaps for a dedicated endpoint it's okay.

What's the difference between that and no-cors, i.e. the tangible security gain for the IDP?

**Comment by annevk** [comment]
@cbiesinger I suspect that "security OK" does rely on a particular implementation whereby responses from the IDP are kept out of the RP process. (Similar to the DoH comment from @martinthomson above. Depending on the characteristics there might be cases where that is okay, though it seems safer to avoid altogether.)

@johannhof it means you cannot attack a non-IDP JSON resource. Either through a request you cannot normally make due to the same-origin policy or being able to read the response through a covert channel.

**Comment by bvandersloot-mozilla** [comment]
> it means you cannot attack a non-IDP JSON resource. Either through a request you cannot normally make due to the same-origin policy or being able to read the response through a covert channel.

@annevk Sorry, but can you elaborate on this? I don't follow.

**Comment by annevk** [comment]
See https://fetch.spec.whatwg.org/#cors-protocol-exceptions, https://educatedguesswork.org/posts/web-security-model-cors/, and https://educatedguesswork.org/posts/web-security-model-side-channels/.

**Comment by martinthomson** [comment]
So I think that this entire CORS issue is bound up in confusion, partly because we have a bunch of very different requests that might be involved.

My understanding is that the accounts endpoint is something that the browser itself hits.  The Origin value in that request would be `null` (or probably absent) in order to maintain our privacy goal (that is, the IdP does not learn about potential RPs, only RPs that the user chose).  The response to that request is consumed by the browser, so there is no cross-site interaction involved.  This request is also likely safe, so there is no inherent need for preflight.  (You might even consider the origin of this request as the IdP itself if that helps.)

The most interesting one is the fetch that a browser makes when a login is initiated.  This is a POST request that results in the IdP generating a token that can be used by the RP.  For this, the IdP probably needs to know the identity of the RP, so Origin can be populated (or the body of the POST can contain the RP identity).  But I don't see any reason why the RP needs to consume the response.  The browser can consume the response and then provide the necessary information to the RP.

This means - in both cases - the IdP doesn't need to concern itself with the various `Access-Control-Allow-...` headers on the response.  Those are for when the browser passes information to another site.  In this case, the IdP can send responses without `Access-Control-Allow-Foo` and be safe in the knowledge that the response will only be usable by the browser.

As for the exceptions, which seem to be largely a preflight exemption, I don't see any reason to exempt FedCM.  There is always a performance angle that argues for preflight exemption, but that really doesn't hold much water here.  Pay the round trip.

**Comment by domfarolino** [comment]
Regarding the accounts endpoint request, my understanding matches @martinthomson's. Honestly, that endpoint seems almost like a `.well-known` endpoint. Maybe it's different since it is supposed to be hit with credentials (do normal `.well-known` requests have credentials included? FWIW one in this spec is currently [using](https://fedidcg.github.io/FedCM/#check-the-root-manifest) no-cors, but maybe that's bad), but the response:
 - Specifically varies on credentials
 - Should not know _who_ is requesting it

That makes CORS feel like an odd property for the request since in order to use credentials it would have to know who is requesting it. Or, yeah I guess it could magically know to always return `null` if the requesting Origin is `null`, as Anne mentioned. That almost feels like a misuse of CORS, but maybe not!

**Comment by martinthomson** [comment]
> do normal .well-known requests have credentials included? 

A `.well-known` endpoint isn't special, at least as far as CORS is concerned.  So the question would have to be answered based on who is asking, where, etc...

**Comment by agektmr** [comment]
I've been innocently believing that fetching FedCM related resources could happen as a first-party, because it's a separate process than the RP's. However, reading through this thread made me believe that treating those requests as third-party / cross-origin makes sense.

I tried to summarize my understanding here.

FedCM sends five different requests:
1. [Top-level manifest](https://developer.chrome.com/blog/fedcm-origin-trial/#top-level-domain-manifest):
  - `GET`
  - `Accept: application/json`
  - uncredentialed.
  - RP shouldn't be disclosed.
2. [FedCM manifest](https://developer.chrome.com/blog/fedcm-origin-trial/#manifest-endpoints):
  - `GET`
  - `Accept: application/json`
  - uncredentialed.
  - RP shouldn't be disclosed.
3. [Client metadata endpoint](https://developer.chrome.com/blog/fedcm-origin-trial/#client-metadata-endpoint):
  - `GET`
  - `Accept: application/json`
  - uncredentialed.
  - **Requires requesting RP's origin.**
4. [Accounts list endpoint](https://developer.chrome.com/blog/fedcm-origin-trial/#accounts-list-endpoint):
  - `GET`
  - `Accept: application/json`
  - **credentialed.**
  - **RP shouldn't be disclosed.**
5. [Assertion endpoint](https://developer.chrome.com/blog/fedcm-origin-trial/#id-token-endpoint):
  - `POST`
  - `Accept: application/x-www-form-urlencoded`
  - **credentialed.**
  - **Requires requesting RP's origin.**

Some premises:
- When we assume the requesting entity (in this case RP?) is cross-origin, 3p cookie must be stripped (or will be stripped in Chrome in the future). FedCM is to give an exception for that.
- FedCM tries to respect the privacy of the user between the RP and the IdP mutually.
- When `Accept` is other than `application/x-www-form-urlencoded`, `multipart/form-data` or `text/plain`, the request must be preflight. This means most of FedCM endpoints must support CORS preflight.
- CORS strips credentials unless explicitly requested.
- CORS involes `Origin` that reveals the requester, but this can be `null` in some condition?

Having said these:

1. Top-level manifest and 2. FedCM manifest can be requested with:
- CORS preflight mode
- Without requesting credentials
- Setting `Origin` as `null`

For 3. Client metadata endpoint
- CORS preflight mode
- **With** requesting credentials
- Setting `Origin` as the RP's origin.

For 4. Accounts list endpoint
- CORS preflight mode
- **With** requesting credentials
- Setting `Origin` as `null`
- Requesting credentials on CORS require an explicit `Access-Control-Allow-Origin` response header with the RP's origin from the IdP, but there is no way for the IdP to learn the origin because it's intentionally hidden.

For 5. Assertion endpoint
- CORS preflight mode
- **With** requesting credentials
- Setting `Origin` as the RP's origin.

I don't think this summary answers questions in this thread but hopefully organizes them better.

**Comment by johannhof** [comment]
> See https://fetch.spec.whatwg.org/#cors-protocol-exceptions, https://educatedguesswork.org/posts/web-security-model-cors/, and https://educatedguesswork.org/posts/web-security-model-side-channels/.

Sorry @annevk, can you be more specific as to how the IDP benefits from using CORS when the only possible response is `null` and without any knowledge on who triggered the request? I agree with @domfarolino that this feels like misuse.

**Comment by npm1** [comment]
I have a proposal to help move this forward. Here is the rationale and proposal:

* These are requests which are initiated by the browser process in Chrome’s implementation, and the responses never reach the RP's renderer process, apart from the final token, once the user has gone through the UI flow.

* The main line of defense of these user-agent-initiated requests is the forbidden header. Whether it be Sec-FedCM-CSRF or some new value for Sec-Fetch-Dest. This is needed because the IDP wants to allow RPs to use FedCM, but not allow them to access the user account data, i.e. not allow an RP-initiated fetch to succeed. From the thread, it sounds like it would be better to use `Sec-Fetch-Dest: web-identity` instead of our own FedCM-specific header.

* I chatted with @domfarolino and he said it would be a good idea to also enforce a response header in addition to the user agent sending a Sec- header in the request. This ensures the IDP is aware of this request being used for FedCM, and acknowledges they've checked for the header. We could use `Support-Loading-Mode: web-identity` or some other header (ideas welcome). 

* For the manifest list, manifest, and account list fetches, we do not want to include a referrer as that would leak the RP to the IDP, which is undesirable. The only option would be to use CORS with `Origin` set to null. It was not clear from this thread what the benefits would be of using null, but if the benefits become clear then we can use CORS with null.

* Client metadata and ID token fetches could use CORS. But this also does not add extra protection on top of what the IDP can achieve by relying on the fact that the fetches include the forbidden header and the response acknowledgement. Because the responses are not shared with the RP at all, except for the ID token, I would say we should add CORS only to the ID token fetches.

I plan to send a PR for these changes soon, hopefully by next week. Let me know if you have any questions or thoughts!

**Comment by cbiesinger** [comment]
For the request in the original title of this issue, I have filed issue #353, since this issue has morphed into general discussions around CORS which is largely orthogonal.

**Comment by cbiesinger** [comment]
OK, we have had extensive discussions on what the best way to proceed is, and here is our proposal.

First, let me summarize the fetches that are done when navigator.credentials.get() is called by the RP:

- The **top level domain manifest** and the **provider manifest** are fetched in parallel. These are uncredentialed fetches with no referrer. The provider manifest is provided by the get() call, while the top level domain manifest points to a well-known file of the ETLD+1 of the provider manifest.
- Once both fetches are finished, we check that the top level domain manifest declares the provider manifest, and if it does not then we fail and do not move on.
- Next, we fetch the **accounts list**: a credentialed request with no referrer. The accounts list url is provided in the provider manifest.
- If the accounts list returns some accounts, we then fetch the **client metadata**, which is also an endpoint provided by the provider manifest. This fetch is uncredentialed but with RP as referrer (note: in the current Chrome implementation we fetch client MD before the accounts list)
- Once the above is finished, we show the FedCM UI to the user if everything is successful so far (ignoring errors from the client MD fetch, if any).
- If the user goes through the UI to sign in with the IDP, we perform the **ID token** fetch, which is credentialed and with RP as referrer.

Second, some context for the proposed solution below:

- We can’t use regular CORS for most endpoints (top level domain manifest, provider manifest, and accounts list), because for privacy reasons we can’t expose the requesting origin
- Requiring CORS for the credentialed endpoints (account list endpoint and the id token endpoint) is undesirable: as long as third-party cookies still exist (I am aware that some browsers have already disabled them), this would make the endpoints readable to the entire web since it would require the server to send `Access-Control-Allow-Origin: null` or `*`. Websites can easily send requests even with a null Origin using a data: URL. (Servers could defend against this by checking Sec-FedCM-XSRF/Sec-Fetch-Dest but since this is a really bad failure case, we’d prefer not to rely on that)
- We could in principle use CORS will a null Origin for manifest fetches and possibly the accounts endpoint, however:
  - Documentation strongly discourages its use (see the note at https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin)
  - A null origin seems harder to understand for IDPs
  - Since websites can send requests with a null origin as described above, this will make the contents world-readable. This could expose e.g. contents of intranet URLs to websites.
- Contrary to some comments, FedCM requests would not trigger a preflight. We do not send any special headers that would require it, and the Accept header (where we use a JSON value) does *not* have an allowlist for MIME types, only the content type header does. (https://fetch.spec.whatwg.org/#cors-safelisted-request-header ).
- For the POST request for the ID token, which sends a Content-Type header, we use the allowlisted application/x-www-form-urlencoded value (based on feedback from web developers, this is easier to use than JSON).
- In principle, you could think of `Access-Control-Allow-Origin: null` on the manifest as an “opt-in” to FedCM. But this is a pretty unspecific way to opt-in. And we already have a much more specific way to opt-in, in the form of the well-known file on the eTLD+1 (e.g. https://google.com/.well-known/web-identity) which needs to list the specific allowed configURL.
- This is largely unrelated, but we did update the spec to be more explicit that endpoint URLs need to be same-origin with the manifest, need to be HTTPS, and that service workers do not get called for FedCM requests.

Given the above, we think CORS adds no value over the `Sec-FedCM-XSRF` ([soon to be changed](https://github.com/whatwg/fetch/pull/1495) to `Sec-Fetch-Dest: webidentity`) header. Concretely, our proposal is:

- Do not use CORS on any of the fetches (see rationale above)
- Do not use `Support-Image-Loading` as the well-known file is sufficient opt-in
- Switch `Sec-FedCM-XSRF` to `Sec-Fetch-Dest: webidentity` to enable IDPs to know when a request comes from the user agent via FedCM.

**Comment by martinthomson** [comment]
This is a pretty comprehensive take, but only from the perspective of the client.  CORS doesn't exist primarily to protect the client; it exists so that servers can be protected from clients.  Clients are allowed to be unhappy about CORS requirements when they are not the beneficiaries of the protections that it provides, but that doesn't change the fact that CORS exists to protect servers that are sometimes unaware of the mechanisms by which they are being protected.

Can you perhaps provide an analysis that supports your conclusion from this perspective?

**Comment by npm1** [comment]
> This is a pretty comprehensive take, but only from the perspective of the client. CORS doesn't exist primarily to protect the client; it exists so that servers can be protected from clients. Clients are allowed to be unhappy about CORS requirements when they are not the beneficiaries of the protections that it provides, but that doesn't change the fact that CORS exists to protect servers that are sometimes unaware of the mechanisms by which they are being protected.
> 
> Can you perhaps provide an analysis that supports your conclusion from this perspective?

From the perspective of protecting servers, the fetches we initiate from FedCM would not trigger preflights if we added CORS. We'd have to actually force preflights to happen, but the fact they do not happen by default means that there are existing mechanisms that can perform similar fetches without preflights. Thus, forcing preflights on FedCM would not introduce any benefits to such servers but it would introduce a lot of issues for clients as mentioned above.

**Comment by cbiesinger** [comment]
From a server protection perspective:

CORS would protect against a) "weird" requests that could break servers, and b) unexpected POST requests that could break servers. In principle it would also protect against requests from unexpected origins, but not here, since we would send a null origin.

However, we already have an opt-in requirement for FedCM -- the well-known file (top-level domain manifest). If it doesn't exist, or if it doesn't list the provider manifest URL, we only make two requests (we make them in parallel in Chrome):
1) A GET request to the tld manifest
2) A GET request to the provider manifest

These are standard GET requests with nothing unusual about them and so they should not affect servers in anyway.

If the well-known file does exist and list the provider manifest, then the server opted in to FedCM and it is safe to make the other requests; CORS would not offer additional protections.

As npm points out, FedCM would not trigger preflight in any case since none of our requests are special enough.

**Comment by bvandersloot-mozilla** [comment]
I agree that for the manifests, CORS with `Origin: null` is probably better as no-CORS.

However I retain reservations about not having CORS on credentialed requests (and see no reason not to use it on Origin-ful requests).

> [...] as long as third-party cookies still exist [...], this would make the endpoints readable to the entire web since it would require the server to send Access-Control-Allow-Origin: null or *. 

Credentialed requests with `Access-Control-Allow-Origin: *` must not have their results shared, per https://fetch.spec.whatwg.org/#cors-protocol-and-credentials. So third parties wouldn't be able to read the results of the requests.

> Websites can easily send requests even with a null Origin using a data: URL. 

Sure, but those requests would fail a [CORS Check](https://fetch.spec.whatwg.org/#ref-for-http-access-control-allow-origin%E2%91%A5). Which means that either the request will have [mode](https://fetch.spec.whatwg.org/#concept-request-mode) "cors" and have a network error or  "no-cors" and the response will be an [opaque filtered response](https://fetch.spec.whatwg.org/#concept-filtered-response-opaque). Tbf, I haven't thought about the "websocket" mode.

**Comment by npm1** [comment]
> 
> Credentialed requests with `Access-Control-Allow-Origin: *` must not have their results shared, per https://fetch.spec.whatwg.org/#cors-protocol-and-credentials. So third parties wouldn't be able to read the results of the requests.

Well this is an IDP we are talking about so they will have many RPs. They can reply with `ACAO: RP`, sure, but that does not remove the concern of RPs reading the data outside of the FedCM API.

> Sure, but those requests would fail a [CORS Check](https://fetch.spec.whatwg.org/#ref-for-http-access-control-allow-origin%E2%91%A5). Which means that either the request will have [mode](https://fetch.spec.whatwg.org/#concept-request-mode) "cors" and have a network error or "no-cors" and the response will be an [opaque filtered response](https://fetch.spec.whatwg.org/#concept-filtered-response-opaque). Tbf, I haven't thought about the "websocket" mode.

They wouldn't fail the CORS check. Null is when the header is not sent, which is different from the string "null". The check will succeed if the Origin value is the string "null" and the origin of the fetch is also the string "null".

**Comment by annevk** [comment]
> The provider manifest is provided by the get() call

Later on in the discussion it's suggested the provider manifest is at a `.well-known` location. Which is it?

Making `.well-known` some kind of same-origin policy extension is an interesting idea, but one that I think would warrant a (much) wider discussion than just this issue/group. In that if you decide to host a resource at a `.well-known` location you opt-in to a different security model. For server administrators that also seems a bit fraught though. Who knows how well the `.well-known` URL space is guarded.

**Comment by npm1** [comment]
> > The provider manifest is provided by the get() call
> 
> Later on in the discussion it's suggested the provider manifest is at a `.well-known` location. Which is it?
> 

Both. `get()` provides the config URL. We fetch the file in the well-known location and check that the config URL specified in that file matches what the `get()` call provided. If it does not, then we reject the `get()` call without performing any additional fetching.

**Comment by bvandersloot-mozilla** [comment]
> They wouldn't fail the CORS check. Null is when the header is not sent, which is different from the string "null". The check will succeed if the Origin value is the string "null" and the origin of the fetch is also the string "null".

Right, 🤦‍♂️, thank you. Good catch. I didn't realize Fetch treats opaque origins identically in this way. Seems not great, but that is why `ACAO: null` is warned against I suppose.

> Well this is an IDP we are talking about so they will have many RPs. They can reply with `ACAO: RP`, sure, but that does not remove the concern of RPs reading the data outside of the FedCM API.

Fair enough. It just makes more sense to me that if a service has its RPs set up to use CORS properly it would manage to validate Sec-Fetch-Dest as well. But you are right that it is a foot-gun.

I'll defer to colleagues with more direct involvement in the Fetch spec as to whether this is an appropriate use case for "no-cors".

**Comment by annevk** [comment]
> (i.e. we only end up fetching the file from .well-known and the config URL)

Okay, so the attacker-supplied URL is always fetched?

**Comment by annevk** [comment]
Thanks Sam!

As alluded to above I can see allowing "`no-cors`" fetches to `.well-known` URLs, although there too I think there's a big risk in that as we do more of that we might forget about the security lessons we learned elsewhere. E.g., if those fetches get attacker-controlled headers, do we ensure the header values cannot be used for attacks or will we forget? As such, I would say that needs a much wider discussion and perhaps it should have dedicated infrastructure in Fetch so it at least has a better chance of getting considered going forward.

Now allowing fetches to arbitrary end points has those risks, but amplified, as the endpoint is even less aware as to what might happen.

Those are the risks with respect to requests. With respect to responses there might also be risks, depending on the implementation strategy, and what we end up exposing over time through various side channels. (E.g., the Web Performance WG always seems keen to expose new numbers to websites and at least historically has not always respected the same-origin policy.)

From all that sticking with the CORS sandbox seems safer.

But that introduces the problem with the accounts fetch. Which is credentialed (cross-site somehow? and I guess that's okay because it's for login) and thus needs protection from random websites fetching it. One idea I had is that we use some kind of user agent origin, e.g., `about:browser`. Websites cannot spoof that. We'd reserve that origin for cases such as this as generally using an opaque origin is preferable per the principle of least privilege.

**Comment by bvandersloot-mozilla** [comment]
> But that introduces the problem with the accounts fetch. Which is credentialed (cross-site somehow? and I guess that's okay because it's for login) and thus needs protection from random websites fetching it. One idea I had is that we use some kind of user agent origin, e.g., about:browser. Websites cannot spoof that. We'd reserve that origin for cases such as this as generally using an opaque origin is preferable per the principle of least privilege.

Ah, neat! I like this as an approach.

**Comment by npm1** [comment]
> As alluded to above I can see allowing "`no-cors`" fetches to `.well-known` URLs, although there too I think there's a big risk in that as we do more of that we might forget about the security lessons we learned elsewhere. E.g., if those fetches get attacker-controlled headers, do we ensure the header values cannot be used for attacks or will we forget? As such, I would say that needs a much wider discussion and perhaps it should have dedicated infrastructure in Fetch so it at least has a better chance of getting considered going forward.
> 
> Now allowing fetches to arbitrary end points has those risks, but amplified, as the endpoint is even less aware as to what might happen.

Are you suggesting using `cors` mode with `origin: null` (or the `about:browser` suggestion below) for these two fetches? Is the claimed benefit that changing the fetch headers used by the API (they are not attacker-controlled) later on might mean we are introducing new attacks on servers because the fetch would have been preflighted if we used CORS?

> Those are the risks with respect to requests. With respect to responses there might also be risks, depending on the implementation strategy, and what we end up exposing over time through various side channels. (E.g., the Web Performance WG always seems keen to expose new numbers to websites and at least historically has not always respected the same-origin policy.)

If you are thinking about Resource Timing, I don't believe these fetches would be exposed there. They need to be opaque to the site using FedCM. Since the user agent ideally needs to consume these fetches outside of the renderer process, it seems hard for Resource Timing to accidentally expose these.

> But that introduces the problem with the accounts fetch. Which is credentialed (cross-site somehow? and I guess that's okay because it's for login) and thus needs protection from random websites fetching it. One idea I had is that we use some kind of user agent origin, e.g., `about:browser`. Websites cannot spoof that. We'd reserve that origin for cases such as this as generally using an opaque origin is preferable per the principle of least privilege.

That seems OK, although it is not clear to me what benefit this introduces over `Sec-Fetch-Dest`. It might be beneficial in the spec to more easily encode how to treat these fetches, and from what I heard from @domfarolino it sounds like FedCM is not the only API trying to add user-agent-only fetches.

**Comment by domfarolino** [comment]
> Is the claimed benefit that changing the fetch headers used by the API (they are not attacker-controlled) later on might mean we are introducing new attacks on servers because the fetch would have been preflighted if we used CORS?

That's my impression of Anne's comment, but yes I'd also like clarification there.

> That seems OK, although it is not clear to me what benefit this introduces over Sec-Fetch-Dest.

I think with something like `about:browser` we could use CORS and the failure mode goes from failing open to failing closed. The failure mode for `Sec-Fetch-Dest` is fail-open because the onus is on the server to opt-out of the whole identity flow by checking `Sec-Fetch-Dest` and returning some bogus/null response if the value isn't `web-identity`; a server that forgets to check this is now "vulnerable". With CORS, the server has to affirmatively opt-in via ACAO. That's not a good strategy for us if we have to use `null`, but if we get to use `about:browser`, then there's a benefit. **But all of this** only seems beneficial if we're not treating the .well-known file as a global opt-in for the identity flow, which seems to be the world @annevk is envisioning?

> and from what I heard from @domfarolino it sounds like FedCM is not the only API trying to add user-agent-only fetches.

Yeah, attribution reporting is currently doing something like this (https://github.com/WICG/attribution-reporting-api/pull/547) and I think ultimately we're going to have to discuss how to best use Fetch for these kind of UA-initiated requests, and some of the trickier bits like what should `client` and `window` be, and what are the full consequences of setting/not-setting these things. But that discussion should probably happen in whatwg/fetch.

**Comment by samuelgoto** [comment]
> and from what I heard from @domfarolino it sounds like FedCM is not the only API trying to add user-agent-only fetches.

Just for completeness, there is a good amount of web platform APIs that use `.well-known` files. The [Change Password](https://w3c.github.io/webappsec-change-password-url/#semantics) and [First Party Sets](https://github.com/WICG/first-party-sets#:~:text=.well%2Dknown%20file%20check%20on%20all%20domains%20to%20ensure%20authorized%20submissions) specs come to mind (but there is probably a bunch more [here](https://en.wikipedia.org/wiki/Well-known_URI)), do we expect them to also require CORS for their `.well-known` files? Do we feel like we need to make them use CORS too or do we feel like there is a criteria that makes us feel comfortable about those?

**Comment by domfarolino** [comment]
Perhaps more interesting than whether the `.well-known` request itself uses CORS, is the fact that we're "loosening" the security characteristics of all requests that participate in the flow specified by the `.well-known` file. I'm consdering:

> In that if you decide to host a resource at a .well-known location you opt-in to a different security model.

...to read as "opt-in to a different security model" (for all requests/flows related to whatever `.well-known` file we fetch). In other words, is the fact that your server has a particular `.well-known` file enough for us to say that's an opt-in to a different security model.

Regarding the actual `.well-known` fetches themselves, it's unfortunately impossible to tell what kind of requests are being made for the examples you mentioned :( [Change password](https://w3c.github.io/webappsec-change-password-url/) spec is pretty incomplete and doesn't ever construct a request or make a fetch, and FPS doesn't have a spec.

**Comment by cbiesinger** [comment]
So I agree that using "Origin: browser" or something like that has fewer negative properties than "Origin: null", but I still don't see the advantage. I can promise you, Anne, that we will not allow the RP to specify custom headers and that we will keep the top-level domain manifest and the provider manifest as simple fetches.

I disagree that the well-known file is opting in to a different security model. It is the same model -- websites can't read responses, only the browser can.

**Comment by cbiesinger** [comment]
One thing that might be interesting as adding a "safe-no-cors" mode to fetch, which does not do the CORS headers *but* still does the checks for safe method/headers and fails the request when it would otherwise trigger precheck. (on the wire it would still send "no-cors" for the mode)

**Comment by annevk** [comment]
> but I still don't see the advantage.

The advantage is we can use CORS and not have to figure out how this extension to the same-origin policy may or may not make sense. On top of which we'd have to support it in perpetuity and ensure it doesn't break any promises made between people in the past.

Note that "safe-no-cors" on its own is no longer enough. You also need the guarantee that the response is not exposed to the process the website runs in. Again, for perpetuity, etc. (For "no-cors" we're creating https://github.com/annevk/orb and as such it would indeed not be adequate as cross-origin JSON would be blocked at the network layer.)

**Comment by annevk** [comment]
@antosart I'm not sure I understand. It wouldn't help with the manifest fetch and there's at least one other fetch that happens in parallel with the manifest fetch.

**Comment by antosart** [comment]
@annevk but those are uncredentialed simple requests which the RP could anyway issue via fetch without triggering CORS preflights, no?

**Comment by annevk** [comment]
@antosart they could, but they would hit a network error due to https://github.com/annevk/orb. In fact, the specification would also hit this network error if it did not introduce a special mode of sorts. "no-cors" isn't exactly reliable infrastructure.

**Comment by domfarolino** [comment]
So they would still hit the server, but orb would block the response? Or would orb step in and block the request too?

**Comment by annevk** [comment]
They would hit the server. The only risk there is really that specifications sometimes grow in ambition over time as already discussed above.

**Comment by cbiesinger** [comment]
A few of us (@annevk, @domfarolino, @martinthomson, @bvandersloot-mozilla, @npm1, @samuelgoto, me) met today to find a consensus on this topic. This is the sumnmary:

- The spec should use Origin instead of Referer for the requests that need to know the RP.
- Requests that should not know the RP should send no Origin (they should not send Origin: null)
- We will continue to not use CORS
- We should add a new mode to the fetch spec (tentatively called "unsafe-no-cors") for requests like this that go to the browser. This needs to specify the client/environment object in a way that works where there is no window object or the like. It also needs to handle CSP in a reasonable way (ie. not require FedCM to explicitly call out to the CSP spec, as we do today). I am not sure if we came to a conclusion about requests that would need preflight; options are to either fail them or do preflight anyway (note that FedCM does not trigger this today). This mode would, of course, not trigger ORB.
- This mode is defined such that the response is only exposed to the browser process. The browser can selectively decide to expose certain data to content; in this case, the ID token. No need to require CORS for this since everything is mediated by the browser.
- The client metadata request may need special handling since fetch does not normally send an Origin header for GET requests, but this endpoint does need to know the origin. This can be handled by FedCM passing in an explicit Origin header to fetch.

In terms of web exposed changes, the only change would be that IDPs need to check the Origin header instead of Referer, and we would send a different Sec-Fetch-Mode header.

Let me know if I missed anything / got it wrong. @bvandersloot-mozilla has volunteered to draft a fetch spec PR, thank you!

**Comment by npm1** [comment]
That sounds right to me. Regarding preflights, I believe we concluded that FedCM fetches specifically do not require preflights.

**Comment by bvandersloot-mozilla** [comment]
I've created a first draft here: https://github.com/whatwg/fetch/pull/1533.

**Comment by npm1** [comment]
An update based on the breakout that happened today. We aligned on @domfarolino's [proposal](https://docs.google.com/document/d/1CpP9JAuqWi4yivOWQcarIqEyQzVcIxDdc8NA3HMw56I/edit) regarding how to treat the accounts endpoint. It will be treated as same-origin, with a null client, and using only SameSite=None cookies. We also briefly discussed the suggestion to add a new header to our CORS request and aligned against it, see the CORS issue https://github.com/fedidcg/FedCM/issues/428.

**Comment by bvandersloot-mozilla** [comment]
That seems like a reasonable state for this. Sorry I couldn't be at the breakout.

**Comment by cbiesinger** [comment]
Both the original title (the header) and the general CORS discussion have now been resolved, afaict (we use CORS for the ID assertion endpoint now). So, closing this, but feel free to comment if you disagree (although filing a new issue may be better as this one is now very long)

**Comment by wseltzer** [comment]
Discussed at 11 April meeting [notes].
